File: backend/src/const/Event.ts
import { TGameState } from "../types/mod";

export const EventState = {
  // teleport
  onTeleport: false,
  onNextTeleport: false,
  onPrevTeleport: false,

  // farm echoes
  onStartFarmEchoes: false,
  onStopFarmEchoes: false,
  onRefreshEchoes: false,
  isEchoesRefreshed: false,

  // Weapon
  onWeaponAdded: false,

  // role
  onRoleReplaced: false,

  // apply buff
  onApplyBuff: false,

  // Dungeon
  onEnterDungeon: false,

  // UID
  onChangeUID: false,

  // Config
  onSaveConfig: false,
  onLoadConfig: false,
};

// daftar event yang auto-reset
export const oneTimeEvents: (keyof typeof EventState)[] = [
  "onTeleport",
  "onNextTeleport",
  "onPrevTeleport",
  "onStartFarmEchoes",
  "onStopFarmEchoes",
  "onApplyBuff",
  "onSaveConfig",
  "onLoadConfig",
  "onEnterDungeon",
  "onWeaponAdded",
  "onRoleReplaced",
  "onChangeUID",
];

File: backend-go/src/models/Event.go
package models

import "github.com/go-playground/validator/v10"

type GetEvent struct {
	OnTeleport        bool `json:"onTeleport"`
	OnNextTeleport    bool `json:"onNextTeleport"`
	OnPrevTeleport    bool `json:"onPrevTeleport"`
	OnStartFarmEchoes bool `json:"onStartFarmEchoes"`
	OnStopFarmEchoes  bool `json:"onStopFarmEchoes"`
	OnSaveConfig      bool `json:"onSaveConfig"`
	OnLoadConfig      bool `json:"onLoadConfig"`
	OnApplyBuff       bool `json:"onApplyBuff"`
	OnEnterDungeon    bool `json:"onEnterDungeon"`
	OnRefreshEchoes   bool `json:"onRefreshEchoes"`
	IsEchoesRefreshed bool `json:"isEchoesRefreshed"`
	OnWeaponAdded     bool `json:"onWeaponAdded"`
	OnRoleReplaced    bool `json:"onRoleReplaced"`
}

type EventUpdate struct {
	OnTeleport        *bool `json:"onTeleport,omitempty"`
	OnNextTeleport    *bool `json:"onNextTeleport,omitempty"`
	OnPrevTeleport    *bool `json:"onPrevTeleport,omitempty"`
	OnStartFarmEchoes *bool `json:"onStartFarmEchoes,omitempty"`
	OnStopFarmEchoes  *bool `json:"onStopFarmEchoes,omitempty"`
	OnSaveConfig      *bool `json:"onSaveConfig,omitempty"`
	OnLoadConfig      *bool `json:"onLoadConfig,omitempty"`
	OnApplyBuff       *bool `json:"onApplyBuff,omitempty"`
	OnEnterDungeon    *bool `json:"onEnterDungeon,omitempty"`
	OnRefreshEchoes   *bool `json:"onRefreshEchoes,omitempty"`
	IsEchoesRefreshed *bool `json:"isEchoesRefreshed,omitempty"`
	OnWeaponAdded     *bool `json:"onWeaponAdded,omitempty"`
	OnRoleReplaced    *bool `json:"onRoleReplaced,omitempty"`
}

var validate = validator.New()

func (e *GetEvent) Validate() error {
	return validate.Struct(e)
}
